lib/vm/src/instance.rs

```diff
@@ -72,45 +72,38 @@ where
             },
         };
         let wasmer_instance = module.instantiate(&import_obj).context(WasmerErr {})?;
-        Ok(Instance::from_wasmer(wasmer_instance, deps))
+        Ok(Instance::from_wasmer(wasmer_instance, deps, gas_limit))
     }
 
-    pub fn from_wasmer(wasmer_instance: wasmer_runtime_core::Instance, deps: Extern<S, A>) -> Self {
-        let res = Instance {
+    pub fn from_wasmer(
+        mut wasmer_instance: wasmer_runtime_core::Instance,
+        deps: Extern<S, A>,
+        gas_limit: u64,
+    ) -> Self {
+        set_gas(&mut wasmer_instance, gas_limit);
+        leave_storage(wasmer_instance.context(), Some(deps.storage));
+        Instance {
             wasmer_instance: wasmer_instance,
             api: deps.api,
             type_storage: PhantomData::<S> {},
-        };
-        res.leave_storage(Some(deps.storage));
-        res
+        }
     }
 
     /// Takes ownership of instance and decomposes it into its components.
     /// The components we want to preserve are returned, the rest is dropped.
-    pub fn recycle(instance: Self) -> (wasmer_runtime_core::Instance, A) {
-        (instance.wasmer_instance, instance.api)
+    pub fn recycle(instance: Self) -> (wasmer_runtime_core::Instance, A, Option<S>) {
+        let storage = take_storage(instance.wasmer_instance.context());
```

Comment:
What about making this Option<Extern>.

If storage is none we can just drop api. It is a simple, clone able struct, so little is lost

Reply:
I don't understand what you mean. The `storage` variable is of type `Option<S>` already due to the return type of `take_storage` from context.rs.

I was even thinking in the opposite direction: why is this optional? How can this ever be unavailable?

Reply:
2 points. Seems my message was eaten by the github

1. I would make the second result Option &lt; Extern &gt; (seems the rust generic syntax got eaten as an html tag)
2. If we now privatize all external ways of having no Storage set (take_storage), and can reason through all internal cases to prove to ourselves this will never happen (like during with_storage), then we can just return `Extern` instead of `Option<Extern>` and panic if not present (only if we prove to devs, not compiler, this will never happen).

I still stand with my point of returning Extern not Storage there

Reply:
1. makes sense, implemented.

2. I found a case where the instance has no storage anymore:

```rust
    #[test]
    fn has_no_storage_when_nested() {
        // this should fail with the assertion, but not cause a double-free crash (issue #59)
        let instance = mock_instance(&CONTRACT_0_7);
        instance.with_storage(|storage1| {
            instance.with_storage(|storage2| {
                assert!(false, "storage missing, so this must not be executed");
            });
        });
    }
```

But for the purpose of this PR we can leave it as is I think.